docs/public-api.asciidoc

```diff
@@ -0,0 +1,265 @@
+
+You only need to use this method to create custom transactions.
+
+It is important to call <<api-transaction-end>> when the transaction has ended.
+A best practice is to use the transaction in a try-catch-finally block or to use the <<convenient-capture-transaction>> method.
+
+Example:
+
+[source,csharp]
+----
+var transaction = Elastic.Apm.Agent
+        .Api.StartTransaction("MyTransaction", Transaction.TYPE_REQUEST);
+try
+{
+    //application code that is captured as a transaction
+}
+catch (Exception e)
+{
+    transaction.CaptureException(e);
+    throw;
+}
+finally
+{
+    transaction.End();
+}
+----
+
+[float]
+[[api-current-transaction]]
+==== `Transaction CurrentTransaction`
+Returns the currently active transaction.
+See <<api-transaction>> on how to customize the current transaction.
+
+If there is no current transaction,
+this method will return null.
+
+[source,csharp]
+----
+var transaction = Elastic.Apm.Agent.Api.CurrentTransaction;
+----
+
+[float]
+[[convenient-capture-transaction]]
+==== `Transaction CaptureTransaction`
+
+This is a convenient method which starts and ends a transaction and captures unhandled exceptions.
+It has 3 parameters:
+
+* `name`: The name of the transaction
+* `type` The type of the transaction
+*  One of the following types which references the code that you want to capture as a transaction: 
+** `Action`
+** `Action<ITransaction>`
+** `Func<T>`
+** `Func<ITransaction,T>`
+** `Func<Task>`
+** `Func<ITransaction,Task>`
+** `Func<Task<T>>`
+** `Func<ITransaction,Task<T>>`
+
+The following code is the equivalent of previous example with the convenient API. It automatically starts and ends the transaction and reports unhandled exceptions. The `t` parameter gives you access to the `ITransaction` instance which represents the transaction that you just created.
+[source,csharp]
+----
+Elastic.Apm.Agent.Api
+        .CaptureTransaction("TestTransaction", Transaction.TYPE_REQUEST, (t) =>
+{
+   //application code that is captured as a transaction
+});
+----
+
+This API also supports `async` methods with the `Func<Task>` overloads.
+
+NOTE: The duration of the transaction will be the timespan between the first and the last line of the `async` lambda expression.
+
+Example:
+[source,csharp]
+----
+await Elastic.Apm.Agent.Api
+        .CaptureTransaction("TestTransaction", "TestType", async () =>
+{
+    //application code that is captured as a transaction
+    await Task.Delay(500); //sample async code
+});
+----
+
+NOTE: The <<convenient-capture-transaction>> method does not start any `Task` for you, so you have to call it with the `await` keyword if you use one of the overloads that except `Task` or `Task<T>`.
+
+
+//----------------------------
+[float]
+[[api-transaction]]
+=== Transaction API
+//----------------------------
+A transaction is the data captured by an agent representing an event occurring in a monitored service
```

Comment:
I'm going to need to fix this paragraph in the java agent too - it's a bit difficult to understand. Here's what I'm thinking:

A transaction describes an event captured by an Elastic APM agent monitoring a service. Transactions help combine multiple <<api-span,`Spans`>> into logical groups, and they are the first <<api-span,`Span`>> of a service. More information on Transactions and Spans is available in the {apm-overview-ref-v}/apm-data-model.html[APM data model] documentation.

Reply:
Ok, I agree with the text, but I think something is not right here. I can build the docu, but in the generated html I don't see this part. Can you please look at this? Is it maybe this link: `{apm-overview-ref-v}/apm-data-model.html[APM data model] `?

Reply:
This is really weird. The link and paragraph both build fine locally for me. Can you do me a favor and update your local copy of the `elastic/docs` repository and try building again?

Reply:
You are right, after I updated my local `elastic/docs` it works. Thank you! ...it was super strange, because before that, the paragraph just did not appear, but I haven't got any error message or warning.

Reply:
Yeah, this is one of the many reasons we're moving from `asciidoc` to `asciidoctor`. In this instance, the docs repo wasn't updated, so the attribute for `{apm-overview-ref-v}` didn't exist locally. Asciidoc then _silently_ fails when you attempt to build the docs. The failure removes any text involved with the link. In other words, if you ever are missing text from an asciidoc file, it's likely due to an attribute being invalid or missing.

Reply:
Good to know, thanks for the info!